src/logics/ArithLogic.cc

```diff
@@ -0,0 +1,1161 @@
+    , sym_Int_LT(declareFun_NoScoping_Chainable(tk_int_lt, sort_BOOL, {sort_INT, sort_INT}))
+    , sym_Int_GEQ(declareFun_NoScoping_Chainable(tk_int_geq, sort_BOOL, {sort_INT, sort_INT}))
+    , sym_Int_GT(declareFun_NoScoping_Chainable(tk_int_gt, sort_BOOL, {sort_INT, sort_INT}))
+    , sym_Int_ITE(sortToIte[sort_INT])
+    , sym_Int_DISTINCT(sortToDisequality[sort_INT])
+
+    , arithType(arithType)
+{
+    assert(static_cast<int>(logicNames.size()) == logicTypes.size());
+}
+
+bool ArithLogic::hasIntegers() const {
+    switch (arithType) {
+        case ArithType::LIA:
+        case ArithType::NIA:
+        case ArithType::IDL:
+        case ArithType::LIRA:
+        case ArithType::NIRA:
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool ArithLogic::hasReals() const {
+    switch (arithType) {
+        case ArithType::LRA:
+        case ArithType::NRA:
+        case ArithType::RDL:
+        case ArithType::LIRA:
+        case ArithType::NIRA:
+            return true;
+        default:
+            return false;
+    }
+}
+
+SymRef ArithLogic::getPlusForSort(SRef sort) const {
+    assert (sort == getSort_int() or sort == getSort_real());
+    return sort == getSort_int() ? get_sym_Int_PLUS() : get_sym_Real_PLUS();
+}
+
+SymRef ArithLogic::getTimesForSort(SRef sort) const {
+    assert (sort == getSort_int() or sort == getSort_real());
+    return sort == getSort_int() ? get_sym_Int_TIMES() : get_sym_Real_TIMES();
+}
+
+SymRef ArithLogic::getMinusForSort(SRef sort) const {
+    assert (sort == getSort_int() or sort == getSort_real());
+    return sort == getSort_int() ? get_sym_Int_MINUS() : get_sym_Real_MINUS();
+};
+
+PTRef ArithLogic::getZeroForSort(SRef sort) const {
+    assert (sort == getSort_int() or sort == getSort_real());
+    return sort == getSort_int() ? getTerm_IntZero() : getTerm_RealZero();
+}
+
+PTRef ArithLogic::getOneForSort(SRef sort) const {
+    assert (sort == getSort_int() or sort == getSort_real());
+    return sort == getSort_int() ? getTerm_IntOne() : getTerm_RealOne();
+}
+
+PTRef ArithLogic::getMinusOneForSort(SRef sort) const {
+    assert (sort == getSort_int() or sort == getSort_real());
+    return sort == getSort_int() ? getTerm_IntMinusOne() : getTerm_RealMinusOne();
+}
+
+bool ArithLogic::isLinearFactor(PTRef tr) const {
+    if (isNumConst(tr) || isNumVarLike(tr)) { return true; }
+    if (isTimes(tr)) {
+        Pterm const& term = getPterm(tr);
+        return term.size() == 2 && ((isNumConst(term[0]) && (isNumVarLike(term[1])))
+                                    || (isNumConst(term[1]) && (isNumVarLike(term[0]))));
+    }
+    return false;
+}
+
+bool ArithLogic::isLinearTerm(PTRef tr) const {
+    if (isLinearFactor(tr)) { return true; }
+    if (isPlus(tr)) {
+        Pterm const& term = getPterm(tr);
+        for (PTRef tr : term) {
+            if (!isLinearFactor(tr)) { return false; }
+        }
+        return true;
```

Comment:
Would `std::all_of` be more expressive here?

Reply:
And pararallelizable! :)

I'll change it.